cognito-post-authentication/src/main/java/no/unit/nva/cognito/service/UserApiClient.java

```diff
@@ -0,0 +1,120 @@
+package no.unit.nva.cognito.service;
+
+import static nva.commons.utils.attempt.Try.attempt;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.net.http.HttpResponse.BodyHandlers;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import no.unit.nva.cognito.model.User;
+import nva.commons.utils.Environment;
+import nva.commons.utils.JacocoGenerated;
+import nva.commons.utils.attempt.ConsumerWithException;
+import nva.commons.utils.attempt.Failure;
+import nva.commons.utils.attempt.Try;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.utils.URIBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UserApiClient implements UserApi {
+
+    public static final String PATH = "/users/";
+    public static final String USER_API_SCHEME = "USER_API_SCHEME";
+    public static final String USER_API_HOST = "USER_API_HOST";
+
+    private final ObjectMapper objectMapper;
+    private final HttpClient httpClient;
+    private final String userApiScheme;
+    private final String userApiHost;
+
+    private static final Logger logger = LoggerFactory.getLogger(UserApiClient.class);
+
+    /**
+     * Constructor for UserApiClient.
+     *
+     * @param httpClient   httpClient
+     * @param objectMapper objectMapper
+     * @param environment  environment
+     */
+    public UserApiClient(HttpClient httpClient, ObjectMapper objectMapper, Environment environment) {
+        this.httpClient = httpClient;
+        this.objectMapper = objectMapper;
+        this.userApiScheme = environment.readEnv(USER_API_SCHEME);
+        this.userApiHost = environment.readEnv(USER_API_HOST);
+    }
+
+    @Override
+    public Optional<User> getUser(String username) {
```

Comment:
A function that returns a function becomes a bit complicated. Here you can avoid it. I have kept the lambda expressions for clarity but feel free to transform them to colon expressions. Also slf4j allows us to include the exception in the log message

```
 @Override
    public Optional<User> getUser(String username) {
        logger.info("Requesting user information for username: " + username);
        return fetchUserInformation(username)
            .stream()
            .filter(responseIsSuccessful())
            .map(this::tryParsingCustomer)
            .collect(SingletonCollector.collect())
            .toOptional(failure -> logErrorParsingUserInformation(failure));
    }

    @Override
    @JacocoGenerated
    public void createUser(User user) {
    }

    private Optional<HttpResponse<String>> fetchUserInformation(String orgNumber) {
        return attempt(()->formUri(orgNumber))
            .map(URIBuilder::build)
            .map(this::buildHttpRequest)
            .map(this::sendHttpRequest)
            .toOptional(failure -> logResponseError(failure));
    }

    private Try<User> tryParsingCustomer(HttpResponse<String> response) {
        return attempt(()->parseUser(response));
    }

    private Predicate<HttpResponse<String>> responseIsSuccessful() {
        return resp -> resp.statusCode() == HttpStatus.SC_OK;
    }
    
    private void logErrorParsingUserInformation(Failure<User> failure) {
        logger.error("Error parsing user information",failure.getException());
    }

    private void logResponseError(Failure<HttpResponse<String>> failure) {
        logger.error("Error fetching user information",failure.getException());
    }

```